### DISCUSSION TITLE
Adding customized paged counter & header

### DISCUSSION THREAD

**Discussion by prunus-cerasus** [sile-typesetter/sile#2115] [Q&A]
I have gone through the entire "SILE Book", looking for the solution, but I guess I'm too stupid.
The issue: I want to change the footer with page counter from "Page number" to "\<Page number\> of \<Total amount of pages\>".
Is it possible? Also is there any lua code to reference these 2 value & more simple way to do it?
The second question is, is there a way to create a header that will be present on top of every one of the pages? e.g. with the name of the author or the document's (e.g. book's) name.

Thank you in advance for the answers.

**Comment by Omikhleia** [comment]
I am converting your issue to a question.

Both are doable, yes, with different possible techniques. My answer below is just one of many alternatives (= what I've done personally for such things!).

At the core, it will require you to write your own document class (derived from an existing one), and/or some small support packages. They would be the most suitable place to implement such hooks.

For the "page of total", you need to track the last page reached at document finish...
  - But as this cannot be known in advance, obviously... so you'd had to store it in some external file at finish...
  - And upon a SILE re-rerun (as in the case of a table of contents, I'm afraid we can't avoid multiple runs here), at initialization you'd have to read that external file, and use its content. (On the first run, the file wouldn't exist obviously, so you'd have to use a placeholder)
  - That's the basic low-level technique (quickly) explained, but note that here a 3rd-party package such as [labelrefs.sile](https://github.com/Omikhleia/labelrefs.sile) may make it simpler: It actually does that exact kind of low level things, i.e. reading a cross-reference files at initialization, saving it at finish, and tracking things at page ends... This is what I used in my CV/resume show case #1272, with pages "1 / N" as seen on the screenshot)[^resume], so my small/naive resume class just has to track a label.[^hammer]

For custom headers (the easiest of the two), you could look at how the default "book" class does it for chapter/section titles, and start from that. The trick here is that such classes ("book" here, but it would also be the case for my own "resilient.book") already have their own page header logic (for showing the chapter/section titles as headers). So you'd likely have to override things for the default page header logic not to apply. Here again, the feasibility can be seen in my resume example (where all pages but the first repeat the candidate's name)[^header] 

So it's doable, yes, but not it's not straightforward, however :)

[^resume]: The code for that small/naive resume class is now part of the [**resilient.sile**](https://github.com/Omikhleia/resilient.sile) collection.

[^header]: And to be honest, there I also re-used my own "resilient.headers" helper package part of my collection of modules, in order not to re-implement again and again the same kind of things...

[^hammer]: **EDIT**: Admittedly, using a cross-reference label for just that purpose is like using a sledgehammer to crack a nut (or "a hammer to crush a fly"), but heh. It does the trick.

**Reply by Omikhleia** [reply]
Regarding a "header that will be present on top of every one of the pages", additionally: It seems to me that if you'd be using the resilient.book class from my above mentioned 3rd-pary [collection](https://github.com/Omikhleia/resilient.sile), you'd have it more or less out-of-the box. I haven't checked right now, but theoretically the following should work (in SIL syntax).[^1]

```
\begin[class=resilient.book, headers=none]{document}
\even-running-header{My Name}
\odd-running-header{My Name}
....
```

[^1]: The `headers=none` class option disables the running header logic (The default value is `technical` where chapter and section titles go in the even and odd running headers, respectively; the other value `novel` would have chapter titles in the odd running headers (and normally the title of the book, if set, in the even running headers); but in your case we wouldn't want any of these). Then we'd explicitly invoke the class low-level commands to set the even- and odd-page running headers.

**Reply by Omikhleia** [reply]
_An additional note;_
My suggestions above would work in the "naive" cases.
But what to expect for works having several high-level constituents, such as a front matter (folios are usually in roman here), etc. In other terms, while everything is achievable at low-level, what is the general solution with a bigger scope in mind? Suddenly it might become less obvious! This remark is probably not useful for @KamilCherry 's initial question, but the answer would then become "Is this doable? Yes! But as a general "standard" solution (= read available "out of the box")? Oh oh... Not so sure regarding the expectations -- Does, say, "xvi of 167" make any sense? :)

**Reply by Omikhleia** [reply]
@KamilCherry Did you have any chance trying one of the above suggestions?
